java/jscip/MessageHandler.java

```diff
@@ -0,0 +1,113 @@
+{
+   private SWIGTYPE_p_SCIP_Messagehdlr _messagehdlrptr; /** pointer address class created by SWIG */
+
+   /** default constructor to construct a Java message handler */
+   protected MessageHandler()
+   {
+      ObjMessagehdlr objmessagehdlr = new ObjMessagehdlr(1) {
+         @Override
+         public void scip_error(SWIGTYPE_p_SCIP_Messagehdlr messagehdlr, SWIGTYPE_p_FILE file, String msg) {
+            error(msg);
+         }
+
+         @Override
+         public void scip_warning(SWIGTYPE_p_SCIP_Messagehdlr messagehdlr, SWIGTYPE_p_FILE file, String msg) {
+            warning(msg);
+         }
+
+         @Override
+         public void scip_dialog(SWIGTYPE_p_SCIP_Messagehdlr messagehdlr, SWIGTYPE_p_FILE file, String msg) {
+            dialog(msg);
+         }
+
+         @Override
+         public void scip_info(SWIGTYPE_p_SCIP_Messagehdlr messagehdlr, SWIGTYPE_p_FILE file, String msg) {
+            info(msg);
+         }
+
+         @Override
+         public SCIP_Retcode scip_free(SWIGTYPE_p_SCIP_Messagehdlr messagehdlr) {
+            try {
+               close();
+               return SCIP_Retcode.SCIP_OKAY;
+            } catch (Exception e) {
+               return SCIP_Retcode.SCIP_ERROR;
+            }
+         }
+      };
+      objmessagehdlr.swigReleaseOwnership();
+      _messagehdlrptr = SCIPJNI.createObjMessagehdlr(objmessagehdlr, 1);
+   }
+
+   /** constructor to wrap a native message handler
+    *  (must not be called directly, construct a Wrapper instead) */
+   private MessageHandler(SWIGTYPE_p_SCIP_Messagehdlr messagehdlrptr)
+   {
+      _messagehdlrptr = messagehdlrptr;
+   }
+
+   /** returns SWIG object type representing a SCIP_Messagehdlr pointer */
+   public SWIGTYPE_p_SCIP_Messagehdlr getPtr()
+   {
+      return _messagehdlrptr;
+   }
+
+   /** returns SWIG object type representing a SCIP_Messagehdlr pointer */
+   public static SWIGTYPE_p_SCIP_Messagehdlr getPtr(MessageHandler obj) {
+      return obj != null ? obj.getPtr() : new SWIGTYPE_p_SCIP_Messagehdlr();
+   }
+
+   protected abstract void error(String msg);
+
+   protected abstract void warning(String msg);
+
+   protected abstract void dialog(String msg);
+
+   protected abstract void info(String msg);
+
+   protected void close() throws Exception {
+      // do nothing by default
+   }
+
+   /** returns a String representation */
+   public String toString()
+   {
+      return "message handler of type " + getClass().getName();
+   }
+
+   /** Class wrapping an existing native SCIP_Messagehdlr. */
+   public static final class Wrapper extends MessageHandler {
+      /** constructor from a native SCIP_Messagehdlr pointer */
+      public Wrapper(SWIGTYPE_p_SCIP_Messagehdlr messagehdlrptr) {
+         super(messagehdlrptr);
+      }
+
+      @Override
+      protected final void error(String msg) {
+         throw new UnsupportedOperationException(
+               "MessageHandler.Wrapper.error should never be called");
```

Comment:
Can you quickly explain why these methods should never be called?

Reply:
From the theoretical standpoint, because calling them will not have the intended effect, as they do not forward to the native message handler's implementation, so they would do nothing.

From a practical standpoint, because those methods are `protected` and the class is `final` (i.e., cannot be subclassed any further), so there is no way they can be called from outside the `jscip` package. (The package thing is unfortunately a loophole that Java leaves here, in C++, it would be completely impossible to call these protected methods of a `final` class from outside the class – note that C++ has supported `final` since C++11.)

Note that the message handler is actually implemented in native C or C++ code and never sees these overrides. To see the overrides, it would 1. have to be a C++ `scip::ObjMessagehdlr` (which at least the default message handler is not), and 2. be wrapped by a subclass of an `ObjMessagehdlr` Java class, so that the SWIG directors feature can do its magic (see, e.g., the class on which you made the previous remark). So there is no way the native code can possibly call back into these methods, either.

(Now if you know a good way to formulate this in fewer words, I can add a comment.)

Reply:
Thanks for the explanation.  Maybe you can just add a short comment "This method will never be reached from native C/C++ code."

Reply:
Done in commit a1aee50f31613852e804709b027599d405567fe2.